### ISSUE TITLE
Cannot read from AWS instance IAM role for AWS provider (as of 5a15c02cbbea27d3f8345b5fe0f348a08a24fdb9)

### ISSUE THREAD

**Issue by jszwedko** [hashicorp/terraform#2828]
It looks like the ability to read from an IAM instance role was lost for the AWS provider. This works fine in 0.6.1, but on the current `HEAD` of master I get:

```
Refreshing Terraform state prior to plan...

Error refreshing state: 1 error(s) occurred:

* 1 error(s) occurred:

* ValidationError: Must specify userName when calling with non-User credentials
        status code: 400, request id: [c3a9805f-30fe-11e5-b474-9f5062e1342d]
```

**Comment by jszwedko** [comment]
I think this may be related to #2730 cc/ @catsby

**Comment by catsby** [comment]
Hey @jszwedko so, are the creds you're using are IAM creds then?

**Comment by catsby** [comment]
Hey @jszwedko yeah, so I made an IAM user and tested to confirm. I added the right privileges and the error went away... 

@phinze thoughts on what to do here? #2730 attempts to simply fail earlier in the process, not sure if I should just revoke this or maybe trap a specific error here :/

**Comment by jszwedko** [comment]
@catsby aaah I see, so one fix would be to give the IAM role access to the IAM service.

**Comment by catsby** [comment]
> aaah I see, so one fix would be to give the IAM role access to the IAM service.

@jszwedko can you confirm that worked for you?

**Comment by catsby** [comment]
Hey @jszwedko – https://github.com/hashicorp/terraform/pull/2959 should address this too

**Comment by jszwedko** [comment]
@catsby awesome, thank you. It turned out the IAM instance profile I was using already had permissions to hit the IAM service. This issue persists in the now released `0.6.2`.

**Comment by catsby** [comment]
yup, that's the then `userName -> ValidationError` thing when using a profile. Fixed in #2959 , thanks for following up!
